src/dispatch/RetryHttpHandler.ts

```diff
@@ -30,14 +31,21 @@ export class RetryHttpHandler implements HttpHandler {
         while (true) {
             try {
                 const response = await this.handler.handle(request);
-                if (this.isStatusCode2xx(response.response.statusCode)) {
+                if (is2xx(response.response.statusCode)) {
                     return response;
                 }
-                throw new Error(`${response.response.statusCode}`);
+                throw response.response.statusCode;
             } catch (e) {
-                if (!retriesLeft) {
+                if (typeof e === 'number' && !is429(e) && !is5xx(e)) {
+                    // Fail immediately on client errors because they will never succeed.
+                    // Only retry when request is throttled (429) or received server error (5xx).
+                    throw new Error(`${e}`);
+                }
+
+                if (retriesLeft <= 0) {
```

Comment:
The purpose is to distinguish the status codes so it only retries on 429 and 5xx; and ignores 400, 403, etc. 

I threw the status code instead of the error because it's easier for `is5xx()` to check a number than a string.